src/shims/fs.rs

```diff
@@ -440,19 +500,68 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     }
 }
 
-// Extracts the number of seconds and nanoseconds elapsed between `time` and the unix epoch, and
-// then sets the `mask` bits determined by `flag` when `time` is Ok. If `time` is an error, it
-// returns `(0, 0)` without setting any bits.
+/// Extracts the number of seconds and nanoseconds elapsed between `time` and the unix epoch when
+/// `time` is Ok. Returns `None` if `time` is an error. Fails if `time` happens before the unix
+/// epoch.
 fn extract_sec_and_nsec<'tcx>(
-    time: std::io::Result<SystemTime>,
-    mask: &mut u32,
-    flag: u32,
-) -> InterpResult<'tcx, (u64, u32)> {
-    if let Ok(time) = time {
+    time: std::io::Result<SystemTime>
+) -> InterpResult<'tcx, Option<(u64, u32)>> {
+    time.ok().map(|time| {
         let duration = system_time_to_duration(&time)?;
-        *mask |= flag;
         Ok((duration.as_secs(), duration.subsec_nanos()))
-    } else {
-        Ok((0, 0))
+    }).transpose()
+}
+
+/// Stores a file's metadata in order to avoid code duplication in the different metadata related
+/// shims.
+struct FileMetadata {
+    mode: Scalar<Tag>,
+    size: u64,
+    created: Option<(u64, u32)>,
+    accessed: Option<(u64, u32)>,
+    modified: Option<(u64, u32)>,
+}
+
+impl FileMetadata {
+    fn new<'tcx, 'mir>(
+        ecx: &mut MiriEvalContext<'mir, 'tcx>,
+        path: PathBuf,
+        follow_symlink: bool
+    ) -> InterpResult<'tcx, Option<FileMetadata>> {
+        let metadata = if follow_symlink {
+            // FIXME: metadata for symlinks need testing.
+            std::fs::symlink_metadata(path)
```

Comment:
The logic here is negated: `symlink_metadata` does *not* follow the symlink, it reads the metadata of the symlink itself rather than the symlink target!

Reply:
ugh sorry I got confused with the negations.let me fix it